### PR TITLE
[SPARK-13549] [SQL] Refactor the Optimizer Rule CollapseProject

### PR DESCRIPTION
**Pull request by gatorsmile**
#### What changes were proposed in this pull request?

The PR https://github.com/apache/spark/pull/10541 changed the rule `CollapseProject` by enabling collapsing `Project` into `Aggregate`. It leaves a to-do item to remove the duplicate code. This PR is to finish this to-do item. Also added a test case for covering this change.
#### How was this patch tested?

Added a new test case. 

@liancheng Could you check if the code refactoring is fine? Thanks!

### CODE REVIEW

**Comment by frreiss** [comment]
LGTM

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -404,69 +404,57 @@ object ColumnPruning extends Rule[LogicalPlan] {
...
+    val aliasMap = AttributeMap(projectList2.collect {
+      case a: Alias => (a.toAttribute, a)
+    })
+
+    // We only collapse these two Projects if their overlapped expressions are all
+    // deterministic.
+    projectList1.exists(_.collect {
+      case a: Attribute if aliasMap.contains(a) => aliasMap(a).child
+    }.exists(!_.deterministic))
+  }
```
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Further refactored your code a little bit:

```scala
def collectAliases(projectList: Seq[NamedExpression]): AttributeMap[Alias] = {
  AttributeMap(projectList.collect {
    case a: Alias => a.toAttribute -> a
  })
}

def haveCommonNonDeterministicOutput(
    upper: Seq[NamedExpression], lower: Seq[NamedExpression]): Boolean = {
  val aliases = collectAliases(lower)
  upper.exists(_.collect {
    case a: Attribute if aliases.contains(a) => aliases(a).child
  }).exists(!_.deterministic)
}

def buildCleanedProjectList(
    upper: Seq[NamedExpression],
    lower: Seq[NamedExpression]): Seq[NamedExpression] = {
  val aliases = collectAliases(lower)

  val rewrittenUpper = upper.map(_.transform {
    case a: Attribute => aliases.getOrElse(a, a)
  })

  rewrittenUpper.map { p =>
    CleanupAliases.trimNonTopLevelAliases(p).asInstanceOf[NamedExpression]
  }
}
```

And those inline comments need some rewording as they are now moved to different contexts.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CollapseProjectSuite.scala [hunk 2]
```diff
@@ -95,4 +95,17 @@ class CollapseProjectSuite extends PlanTest {
...
+      .groupBy('a, 'b)(('a + 1).as('a_plus_1), 'b)
+      .select('a_plus_1, ('b + 1).as('b_plus_1))
+
+    val optimized = Optimize.execute(query.analyze)
+
+    val correctAnswer = testRelation
+      .groupBy('a, 'b)(('a + 1).as('a_plus_1), ('b + 1).as('b_plus_1)).analyze
+
+    comparePlans(optimized, correctAnswer)
+  }
```
Comment (liancheng) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CollapseProjectSuite.scala]: Would be nice to add one more test case that contains common non-deterministic fields.

**Comment by liancheng** [comment]
Sorry for the late review. In general it looks good. Left some comments for better readability and testing.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CollapseProjectSuite.scala (hunk 2 above)
Comment (gatorsmile) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CollapseProjectSuite.scala]: Sure, will do it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 1 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Great! will do

**Comment by gatorsmile** [comment]
This is definitely a low priority issue. Thank you for your review! Will do the change based on your suggestions. @liancheng

**Comment by liancheng** [comment]
Thanks! Merging to master.

**Comment by gatorsmile** [comment]
Thank you! @liancheng 

Now, this rule supports collapsing two continuous `Project` operators and one upper `Project` and one lower `Aggregate`. 

After rethinking it, maybe we also can collapse one upper `Aggregate` into one lower `Project`. Will try to submit a PR for this. : )

**Comment by gatorsmile** [comment]
Let me post an example here:
```SQL
SELECT sum(b) FROM parquet_t2 group by a, b
```
The current optimized plan is like
```
Aggregate [a#43L,b#44L], [(sum(b#44L),mode=Complete,isDistinct=false) AS sum(b)#51L]
+- Project [a#43L,b#44L]
   +- Relation[a#43L,b#44L,c#45L,d#46L] ParquetFormat part: struct<>, data: struct<a:bigint,b:bigint,c:bigint,d:bigint>
```

Ideally, we can directly remove the Project.
```
Aggregate [a#43L,b#44L], [(sum(b#44L),mode=Complete,isDistinct=false) AS sum(b)#51L]
 +- Relation[a#43L,b#44L,c#45L,d#46L] ParquetFormat part: struct<>, data: struct<a:bigint,b:bigint,c:bigint,d:bigint>
```

**Comment by gatorsmile** [comment]
But, their physical plan is still the same. I guess, we do not need to do it. : )

**Comment by liancheng** [comment]
@gatorsmile Yea, whole-stage codegen makes this kind of optimization not that appealing.

**Comment by gatorsmile** [comment]
I see. Thanks! @liancheng Let me find something else in Optimizer. Will do expression folding this weekend. : )

**Pull request closed without merging**